### PR TITLE
Rework broadcasting setup to decrease binary size.

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Rework broadcasting setup to decrease binary size. Push all the type specific logic down and separate out the broadcasting/parallelization.

Reductions:
  - element_wise_ops: 521.0 KB -> 268.8 KB (-252.2 KB)
  - qlinear_binary_op: 28.1 -> 12.8 (-15.3 KB)
  - where: 25.8 KB -> 17.3 KB (-8.5 KB)

Total reduction: 276 KB

**Motivation and Context**
Reduce binary size.

### CODE REVIEW

**Comment by skottmckay** [comment]
Perf tested with a model that had 800 nodes (400 Add, 400 Sub) using various input sizes. Performance was equivalent to the existing code.

**Comment by yuslepukhin** [comment]
It looks like, binary size reduction is becoming an art on its own. It may be worth to create a OneNote page with all the best practices so the optimal binary size is achieved on the outset rather than after refactoring.

---
In reply to: [695953799](https://github.com/microsoft/onnxruntime/pull/5227#issuecomment-695953799) [](ancestors = 695953799)

File: onnxruntime/core/providers/cpu/math/element_wise_ops.cc [hunk 1]
```diff
@@ -216,118 +216,151 @@ ONNX_CPU_OPERATOR_KERNEL(
...
-  if (Y.Shape().Size() == 1) {
-    auto exp = *Y.template Data<E>();
-    if (exp == E{2}) {
-      input1scalar = [](gsl::span<T> output, gsl::span<const T> X, E) {
-        std::transform(X.cbegin(), X.cend(), output.begin(),
-                       [](T x) {
-                         return static_cast<T>(x * x);
+        std::transform(Y.cbegin(), Y.cend(), output.begin(),
+                       [X](E y) {
+                         return static_cast<T>(std::pow(X, y));
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/math/element_wise_ops.cc:307]: These are also lambdas. Would rewriting them as a loop help?

**Review by yuslepukhin — APPROVED** [review]
:shipit:

File: onnxruntime/core/providers/cpu/math/element_wise_ops.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/math/element_wise_ops.cc:307]: Lambdas aren't necessarily problematic. In this case std::transform is templatized so the lambda is not being pushed into a std::function and there's no unnecessary code covered by template. Based on that I wouldn't expect rewriting as a loop to provide much if any benefit.

**Comment by snnn** [comment]
Looks like this change caused 30% perf decrease on the bertsquad model.

**Pull request merged**